# Car Price Prediction using Simple Linear Regression

This dataset contains some characteristics of cars and it's a clean dataset that was downlaoded from Kaggle. Every car has a name, unique id, fuel type etc. I will be performing simple linear regression to predict car price

In [1]:
# Importing the require libraries
import numpy as np
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.metrics import r2_score

In [2]:
# Importing the dataset
autos = pd.read_csv(r'C:\Users\Chuks\datasets\CarPrice_Assignment.csv')
autos.head()

car_ID  symboling                   CarName fueltype aspiration doornumber  \
0       1          3        alfa-romero giulia      gas        std        two   
1       2          3       alfa-romero stelvio      gas        std        two   
2       3          1  alfa-romero Quadrifoglio      gas        std        two   
3       4          2               audi 100 ls      gas        std       four   
4       5          2                audi 100ls      gas        std       four   

       carbody drivewheel enginelocation  wheelbase  ...  enginesize  \
0  convertible        rwd          front       88.6  ...         130   
1  convertible        rwd          front       88.6  ...         130   
2    hatchback        rwd          front       94.5  ...         152   
3        sedan        fwd          front       99.8  ...         109   
4        sedan        4wd          front       99.4  ...         136   

   fuelsystem  boreratio  stroke compressionratio horsepower  peakrpm citympg  \
0        mpfi       3.47    2.68              9.0        111     5000      21   
1        mpfi       3.47    2.68              9.0        111     5000      21   
2        mpfi       2.68    3.47              9.0        154     5000      19   
3        mpfi       3.19    3.40             10.0        102     5500      24   
4        mpfi       3.19    3.40              8.0        115     5500      18   

   highwaympg    price  
0          27  13495.0  
1          27  16500.0  
2          26  16500.0  
3          30  13950.0  
4          22  17450.0  

[5 rows x 26 columns]

In [3]:
autos.shape

(205, 26)

In [4]:
autos.columns

Index(['car_ID', 'symboling', 'CarName', 'fueltype', 'aspiration',
       'doornumber', 'carbody', 'drivewheel', 'enginelocation', 'wheelbase',
       'carlength', 'carwidth', 'carheight', 'curbweight', 'enginetype',
       'cylindernumber', 'enginesize', 'fuelsystem', 'boreratio', 'stroke',
       'compressionratio', 'horsepower', 'peakrpm', 'citympg', 'highwaympg',
       'price'],
      dtype='object')

In [5]:
autos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 205 entries, 0 to 204
Data columns (total 26 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   car_ID            205 non-null    int64  
 1   symboling         205 non-null    int64  
 2   CarName           205 non-null    object 
 3   fueltype          205 non-null    object 
 4   aspiration        205 non-null    object 
 5   doornumber        205 non-null    object 
 6   carbody           205 non-null    object 
 7   drivewheel        205 non-null    object 
 8   enginelocation    205 non-null    object 
 9   wheelbase         205 non-null    float64
 10  carlength         205 non-null    float64
 11  carwidth          205 non-null    float64
 12  carheight         205 non-null    float64
 13  curbweight        205 non-null    int64  
 14  enginetype        205 non-null    object 
 15  cylindernumber    205 non-null    object 
 16  enginesize        205 non-null    int64  
 1

In [6]:
null_counts = autos.isnull().sum()
null_counts[null_counts > 0]

Series([], dtype: int64)

The information abvoe show there are 205 records and 26 features in the dataset with no missing or null values. The features are made up of 10 categorical features and 16 numeric feature ( 8 integer and 8 float). Some irrelevant features like car_ID, symboling, CarName will be dropped and the categorical columns will be converted to numeric data using onehot encoding by invoking pd.get_dumies() function

In [7]:
# Droopping a few irrelevant features
autos.drop(['car_ID', 'symboling', 'CarName'], axis=1, inplace=True)

In [8]:
autos = pd.get_dummies(autos)

In [9]:
autos.head()

wheelbase  carlength  carwidth  carheight  curbweight  enginesize  \
0       88.6      168.8      64.1       48.8        2548         130   
1       88.6      168.8      64.1       48.8        2548         130   
2       94.5      171.2      65.5       52.4        2823         152   
3       99.8      176.6      66.2       54.3        2337         109   
4       99.4      176.6      66.4       54.3        2824         136   

   boreratio  stroke  compressionratio  horsepower  ...  \
0       3.47    2.68               9.0         111  ...   
1       3.47    2.68               9.0         111  ...   
2       2.68    3.47               9.0         154  ...   
3       3.19    3.40              10.0         102  ...   
4       3.19    3.40               8.0         115  ...   

   cylindernumber_twelve  cylindernumber_two  fuelsystem_1bbl  \
0                      0                   0                0   
1                      0                   0                0   
2                      0                   0                0   
3                      0                   0                0   
4                      0                   0                0   

   fuelsystem_2bbl  fuelsystem_4bbl  fuelsystem_idi  fuelsystem_mfi  \
0                0                0               0               0   
1                0                0               0               0   
2                0                0               0               0   
3                0                0               0               0   
4                0                0               0               0   

   fuelsystem_mpfi  fuelsystem_spdi  fuelsystem_spfi  
0                1                0                0  
1                1                0                0  
2                1                0                0  
3                1                0                0  
4                1                0                0  

[5 rows x 52 columns]

In [10]:
autos.shape

(205, 52)

In [11]:
x = autos.drop('price', axis=1)
y = autos['price']

In [12]:
x_train, x_test, y_train, y_test = train_test_split(x, y, test_size=0.2)

In [13]:
x_train.shape, x_test.shape, y_train.shape, y_test.shape

((164, 51), (41, 51), (164,), (41,))

So far after invoking pd.get_dumies() function on the dataset the columns features increased to 52 from 25. Since we want to predict the price we make it the target ie the dependent variable while all other feature are labels or indeopendent varibales. Splitting the dataset into training and testing data (80% and 20% respectively) to be used for training and testing our model. This produces a training data with 164 records and testing data with 41 records

We will go ahead to istantiate the linear reression model, fit the training data and see what score the model will produce.

In [14]:
model = LinearRegression().fit(x_train, y_train)

In [15]:
Training_score = model.score(x_train, y_train)

Training_score

0.9426608584134227

In [16]:
y_pred = model.predict(x_test)

y_pred

array([ 6748.51862491,  9682.42069382, -1043.57501373, 11151.21702643,
       43911.07883024, 40340.60404951, 17499.75443669, 10765.20347231,
        8252.39656457, 19220.89787779,  7087.27917296,  9798.27372129,
       20318.62414493, 10581.05012428, 10023.53328035,  9833.37780645,
        4519.97712337,  8580.9420425 , 17427.76662011, 16178.16284564,
        5258.72212945, 11506.40351532, 20847.86612428, 17141.27018211,
        6867.54389549, 24825.77047102,  6320.16557761,  9565.27447658,
        6716.43871935,  6663.09587064,  6545.83713258, 13224.76728048,
        5085.22823446,  6935.86034231, 10353.71035459,  5751.96955636,
        7007.53589933,  7123.92378204,  6910.87455144, 11604.8428399 ,
       12307.87317802])

In [17]:
Testing_score = r2_score(y_test, y_pred)

Testing_score

0.8281859857795324

In [18]:
model.coef_

array([-3.98032116e+01, -2.94561222e+01,  6.81799579e+02,  1.56846736e+02,
        4.00919594e+00,  1.30816979e+02, -1.35538344e+03, -4.82003085e+03,
       -6.85423869e+02,  1.12625023e+01,  2.42464122e+00, -1.27432526e+02,
        2.02432989e+02,  3.05925583e+03, -3.05925583e+03, -1.10736519e+03,
        1.10736519e+03, -5.00104403e+02,  5.00104403e+02,  2.10527753e+03,
       -2.05842463e+03, -6.47067087e+02,  8.95941087e+02, -2.95726897e+02,
       -1.19050541e+02, -2.59686658e+02,  3.78737199e+02, -4.01716727e+03,
        4.01716727e+03, -6.41631983e+01, -1.81898940e-12, -2.42488531e+02,
        3.23935230e+03,  6.27656636e+02, -5.71723137e+03,  2.15687417e+03,
        7.70416292e+03, -1.81049448e+03, -2.57802217e+03,  5.97110987e+02,
        0.00000000e+00, -6.06963142e+03,  2.15687417e+03, -4.37912347e+01,
       -1.12350786e+02,  2.15687417e+03,  3.05925583e+03,  0.00000000e+00,
       -6.76231907e+02, -3.56983435e+03, -8.13921721e+02])

In [19]:
model.intercept_

-40411.21951635902

The model coefficient and intercept are displayed above. R Squared is a goodness of fit for linear regression models. It measures the relationship between the model and the dependent variable on a scale of 0 - 100%

The score of 94% and 79% for training and testing data respectively is a good one and can be relied on. However our model will need to be constantly evaluated as the model accuracy will likely degrade over time. Though R squared has its limitationsbut this will be a topic for another articule